### PR TITLE
Add UserController to RightsService

### PR DESCRIPTION
**Pull request by SolovyovDmitry**
Add UserController to RightsService with edit endpoint which allows to create/edit/remove user roles

### CODE REVIEW

File: src/RightsService.Business/Commands/User/EditUserRoleCommand.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      _responseCreator = responseCreator;
+      _validator = validator;
+      _mapper = mapper;
+      _repository = repository;
+    }
+
+    public async Task<OperationResultResponse<Guid?>> ExecuteAsync(EditUserRoleRequest request)
+    {
+      // TODO: Check proper right
+      if (!await _accessValidator.HasRightsAsync(Rights.AddEditRemoveUsers))
```
Comment (MariyaSin) [inline src/RightsService.Business/Commands/User/EditUserRoleCommand.cs:49]: AddRemoveUsersRoles

File: src/RightsService.Business/Commands/User/EditUserRoleCommand.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    {
+      _accessValidator = accessValidator;
+      _httpContextAccessor = httpContextAccessor;
+      _responseCreator = responseCreator;
+      _validator = validator;
+      _mapper = mapper;
+      _repository = repository;
+    }
+
+    public async Task<OperationResultResponse<Guid?>> ExecuteAsync(EditUserRoleRequest request)
```
Comment (MariyaSin) [inline src/RightsService.Business/Commands/User/EditUserRoleCommand.cs:46]: edit operations returns bool

Comment (MariyaSin) [reply src/RightsService.Business/Commands/User/EditUserRoleCommand.cs:46]: Task<OperationResultResponse<bool>>

File: src/RightsService.Business/Commands/User/EditUserRoleCommand.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+          validationResult.Errors.Select(e => e.ErrorMessage).ToList());
+      }
+
+      OperationResultResponse<Guid?> response = new();
+      await _repository.RemoveAsync(request.UserId);
+      response.Body = request.RoleId.HasValue
+        ? (await _repository.CreateAsync(_mapper.Map(
+          request,
+          _httpContextAccessor.HttpContext.GetUserId())))
+        : null;
```
Comment (MariyaSin) [inline src/RightsService.Business/Commands/User/EditUserRoleCommand.cs:69]: bool

File: src/RightsService.Business/Commands/User/EditUserRoleCommand.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      response.Status = OperationResultStatusType.FullSuccess;
```
Comment (MariyaSin) [inline src/RightsService.Business/Commands/User/EditUserRoleCommand.cs:70]: it depaen on response.Body value

File: src/RightsService.Business/Commands/User/Interfaces/IEditUserRoleCommand.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+using LT.DigitalOffice.Kernel.Attributes;
+using LT.DigitalOffice.Kernel.Responses;
+using LT.DigitalOffice.RightsService.Models.Dto.Requests;
+
+namespace LT.DigitalOffice.RightsService.Business.Commands.User.Interfaces
+{
+  [AutoInject]
+  public interface IEditUserRoleCommand
+  {
+    Task<OperationResultResponse<Guid?>> ExecuteAsync(EditUserRoleRequest request);
```
Comment (MariyaSin) [inline src/RightsService.Business/Commands/User/Interfaces/IEditUserRoleCommand.cs:12]: Guid => bool

File: src/RightsService.Mappers/Db/DbUserRoleMapper.cs [hunk 6]
```diff
@@ -19,5 +20,18 @@ public DbUserRole Map(ICreateUserRolePublish request)
         IsActive = true
       };
     }
+
+    public DbUserRole Map(EditUserRoleRequest request, Guid changedBy)
```
Comment (MariyaSin) [inline src/RightsService.Mappers/Db/DbUserRoleMapper.cs:24]: you can get the changedBy value from HTTP context.

File: src/RightsService.Mappers/Db/DbUserRoleMapper.cs [hunk 7]
```diff
@@ -19,5 +20,18 @@ public DbUserRole Map(ICreateUserRolePublish request)
...
+    {
+      return new DbUserRole
+      {
+        Id = Guid.NewGuid(),
+        UserId = request.UserId,
+        RoleId = request.RoleId.Value,
+        CreatedAtUtc = DateTime.UtcNow,
+        CreatedBy = changedBy,
```
Comment (MariyaSin) [inline src/RightsService.Mappers/Db/DbUserRoleMapper.cs:32]: _httpContextAccessor.HttpContext.GetUserId(),

File: src/RightsService/Controllers/UserController.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+using Microsoft.AspNetCore.Mvc;
+
+namespace LT.DigitalOffice.RightsService.Controllers
+{
+  [Route("[controller]")]
+  [ApiController]
+  public class UserController : ControllerBase
+  {
+    [HttpPut("edit")]
+    public async Task<OperationResultResponse<Guid?>> EditAsync(
```
Comment (MariyaSin) [inline src/RightsService/Controllers/UserController.cs:15]: bool

File: src/RightsService.Business/Commands/User/EditUserRoleCommand.cs (hunk 1 above)
Comment (SolovyovDmitry) [reply src/RightsService.Business/Commands/User/EditUserRoleCommand.cs:49]: No such right. Shall I update kernel?

File: src/RightsService.Validation/EditUserRoleRequestValidator.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace LT.DigitalOffice.RightsService.Validation
+{
+  public class EditUserRoleRequestValidator : AbstractValidator<EditUserRoleRequest>, IEditUserRoleRequestValidator
+  {
+    public EditUserRoleRequestValidator(
+      IRoleRepository repository,
+      IUserService userService)
+    {
+      RuleFor(x => x.UserId)
```
Comment (MariyaSin) [inline src/RightsService.Validation/EditUserRoleRequestValidator.cs:17]: check if Guid.Empty

File: src/RightsService.Business/Commands/User/EditUserRoleCommand.cs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+      if (!validationResult.IsValid)
+      {
+        return _responseCreator.CreateFailureResponse<bool>(
+          HttpStatusCode.BadRequest,
+          validationResult.Errors.Select(e => e.ErrorMessage).ToList());
+      }
+
+      OperationResultResponse<bool> response = new();
+      bool removalResult = await _repository.RemoveAsync(request.UserId);
```
Comment (MariyaSin) [inline src/RightsService.Business/Commands/User/EditUserRoleCommand.cs:57]: if (!(await _repository.RemoveAsync(request.UserId)) && !requset.RoleId.HasValue) => badRequest

File: src/RightsService.Business/Commands/User/EditUserRoleCommand.cs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+      response.Body = request.RoleId.HasValue
+        ? (await _repository.CreateAsync(_mapper.Map(
+          request))).HasValue
+        : false;
```
Comment (MariyaSin) [inline src/RightsService.Business/Commands/User/EditUserRoleCommand.cs:61]: true

File: src/RightsService.Data/UserRoleRepository.cs [hunk 12]
```diff
@@ -95,17 +95,18 @@ public async Task<List<DbUserRole>> GetWithRightsAsync()
         .ToListAsync();
     }
 
-    public async Task RemoveAsync(Guid userId)
+    public async Task<bool> RemoveAsync(Guid userId)
     {
       DbUserRole user = _provider.UsersRoles.FirstOrDefault(u => u.UserId == userId && u.IsActive);
```
Comment (MariyaSin) [inline src/RightsService.Data/UserRoleRepository.cs:100]: async

File: src/RightsService.Data/UserRoleRepository.cs [hunk 13]
```diff
@@ -95,17 +95,18 @@ public async Task<List<DbUserRole>> GetWithRightsAsync()
...
 
       if (user is null)
       {
-        return;
+        return false;
       }
 
       user.IsActive = false;
```
Comment (MariyaSin) [inline src/RightsService.Data/UserRoleRepository.cs:107]: changedAt, ChangedBy

File: src/RightsService.Business/Commands/User/EditUserRoleCommand.cs [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+      OperationResultResponse<bool> response = new();
+
+      if (!(await _repository.RemoveAsync(request.UserId)) && !request.RoleId.HasValue)
+      {
+        return _responseCreator.CreateFailureResponse<bool>(
+          HttpStatusCode.BadRequest,
+          new List<string> { "Removal failed" });
+      }
+
+      bool removalResult = await _repository.RemoveAsync(request.UserId);
```
Comment (MariyaSin) [inline src/RightsService.Business/Commands/User/EditUserRoleCommand.cs:66]: do you need the variable removalResult?
you can write await _repository.RemoveAsync(request.UserId);

File: src/RightsService.Validation/EditUserRoleRequestValidator.cs [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+        .Must(x => x != Guid.Empty)
+        .WithMessage("No user id provided.");
+
+      RuleFor(x => x.UserId)
```
Comment (MariyaSin) [inline src/RightsService.Validation/EditUserRoleRequestValidator.cs:21]: remove 20 and 21 line to join check of UserId
add Cascade before 17 line

File: src/RightsService.Business/Commands/User/EditUserRoleCommand.cs [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+      OperationResultResponse<bool> response = new();
+
+      if (!(await _repository.RemoveAsync(request.UserId)) && !request.RoleId.HasValue)
+      {
+        return _responseCreator.CreateFailureResponse<bool>(
+          HttpStatusCode.BadRequest,
+          new List<string> { "Removal failed" });
+      }
+
+      await _repository.RemoveAsync(request.UserId);
```
Comment (MariyaSin) [inline src/RightsService.Business/Commands/User/EditUserRoleCommand.cs:66]: remove

File: src/RightsService.Data/UserRoleRepository.cs [hunk 17]
```diff
@@ -95,17 +100,20 @@ public async Task<List<DbUserRole>> GetWithRightsAsync()
...
+      user.ModifiedAtUtc = DateTime.Now;
+      user.ModifiedBy = _httpContextAccessor.HttpContext.GetUserId();
       await _provider.SaveAsync();
```
Comment (ndoriomedov) [inline src/RightsService.Data/UserRoleRepository.cs:115]: add empty line after this

**Review by ndoriomedov — APPROVED** [review]

**Review by MariyaSin — APPROVED** [review]

**Pull request merged**